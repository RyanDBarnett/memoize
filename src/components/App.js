import React, { Component } from 'react';
import '../styles/App.css';
import Header from './Header';
import Card from './Card';
import balloon from '../media/hot-air-balloon.png';

class App extends Component {
  constructor() {
    super();
    this.state = {
      currentQuestions: null,
      currentQuestion: {},
      correctQuestions: [],
      incorrectQuestions: []
    }
  }

  componentDidMount = () => {
    fetch('https://fe-apps.herokuapp.com/api/v1/memoize/1901/ryandbarnett/cards')
    .then(response => response.json())
    .then(data => {
      this.setState({
        currentQuestions: Object.assign([], data.cards).splice(1),
        currentQuestion: Object.assign({}, data.cards[0])
      })
    })
    .catch(error => {throw new Error(error)})
  }

  checkAnswer = (answer) => {
    if(this.state.currentQuestion.answer === answer) {
      this.setState({
        correctQuestions: this.state.correctQuestions.concat([this.state.currentQuestion])
      })
    } else {
      this.setState({
        incorrectQuestions: this.state.incorrectQuestions.concat([this.state.currentQuestion])
      })
    }
  }
  
  showNextQuestion = () => {
    this.setState({
      currentQuestion: this.state.currentQuestions.shift()
    });
    this.setStateLocalStorage();
  }

  resetAllQuestions = () => {
    const allQuestions = this.state.correctQuestions.concat(this.state.incorrectQuestions)
    this.setState({
      currentQuestions: allQuestions.splice(1),
      currentQuestion: allQuestions[0],
      correctQuestions: [],
      incorrectQuestions: []
    })
  }

  resetIncorrectQuestions = () => {
    this.setState({
      currentQuestions: this.state.incorrectQuestions.splice(1),
      currentQuestion: this.state.incorrectQuestions[0],
      incorrectQuestions: []
    })
  }

  setStateLocalStorage = () => {
    localStorage.setItem('appState', JSON.stringify(this.state));
  }

  getStateLocalStorage = () => {
    return JSON.parse(localStorage.getItem('appState'));
  }

  addBalloons = () => {
    return (
      this.state.correctQuestions.map((correctQues, index) => {
        return <img className="hot-air-balloon" key={index} src={balloon} style={{top: this.getRandomWindowHeight()}} alt="Orange and red striped hot air balloon"/>
      })
    )
  }

  getRandomWindowHeight = () => {
    let height = window.innerHeight;
    return Math.floor(Math.random() * height) + 1;
  }

  render() {
    let balloons = this.state.correctQuestions && this.addBalloons();
    let card = this.state.currentQuestions !== null &&
      <Card
        currentQuestion={this.state.currentQuestion}
        showNextQuestion={this.showNextQuestion}
        resetAllQuestions={this.resetAllQuestions}
        resetIncorrectQuestions={this.resetIncorrectQuestions}
        checkAnswer={this.checkAnswer}
      />;

    return (
      <div className="App">
        {balloons}
        <Header />
        {card}
      </div>
    );
  }
}

export default App;
