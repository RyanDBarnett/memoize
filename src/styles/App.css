* {
  font-family: 'OpenDyslexic';
}

.App {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-image: url("../media/chuttersnap-541390-unsplash.jpg");
  background-size: cover;
  background-position: center;
}

.App-header {
  background: #e3f0f6cc;
  margin: 1%;
  padding: 0 4%;
  border-radius: 5%;
  border: #211a90 2px solid;
}

.mandala {
  animation: mandala-spin infinite 120s linear;
  position: absolute;
  width: 40%; 
}

@keyframes mandala-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
