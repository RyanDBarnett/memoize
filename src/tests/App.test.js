import React from 'react';
import ReactDOM from 'react-dom';
import App from '../components/App';
import { shallow } from "enzyme";

describe('App', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <App />
    );
  });

  it("should match the snapshot with all data passed in", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('checkAnswer', () => {
    it('should put the currentQuestion into correctQuestions if answer matches', () => {
      let wrapper2 = shallow(<App />);
      wrapper2.setState({
        currentQuestions: [],
        currentQuestion: {
          answer: "Roles",
          choices: ["Properties","States","Roles"],
          id: "1",
          question: "Which of the 3 ARIA main features defines what an element is or does?"
        },
        correctQuestions: [],
        incorrectQuestions: []
      });
      wrapper2.instance().checkAnswer('Roles');
      expect(wrapper2.state().correctQuestions.length).toEqual(1);
    });

    it('should put the currentQuestion into incorrectQuestions if answer matches', () => {
      let wrapper2 = shallow(<App />);
      wrapper2.setState({
        currentQuestions: [],
        currentQuestion: {
          answer: "Roles",
          choices: ["Properties","States","Roles"],
          id: "1",
          question: "Which of the 3 ARIA main features defines what an element is or does?"
        },
        correctQuestions: [],
        incorrectQuestions: []
      });
      wrapper2.instance().checkAnswer('States');
      expect(wrapper2.state().incorrectQuestions.length).toEqual(1);
    });
  });

  describe('showNextQuestion', () => {
    it('should set the first question in currentQuestions to the currentQuestion', () => {
      let wrapper3 = shallow(<App />);
      wrapper3.setState({
        currentQuestions: [
          {
            answer: "Roles",
            choices: ["Properties","States","Roles"],
            id: "1",
            question: "Which of the 3 ARIA main features defines what an element is or does?"
          }
        ],
        currentQuestion: null,
      });
      expect(wrapper3.state().currentQuestions.length).toEqual(1);
      expect(wrapper3.state().currentQuestion).toEqual(null)
      wrapper3.instance().showNextQuestion();
      expect(wrapper3.state().currentQuestions.length).toEqual(0);
      expect(wrapper3.state().currentQuestion).toEqual({
        answer: "Roles",
        choices: ["Properties","States","Roles"],
        id: "1",
        question: "Which of the 3 ARIA main features defines what an element is or does?"
      });
    });
  });

  describe('resetAllQuestions', () => {
    beforeEach( () => {
      wrapper.setState({
        currentQuestions: [],
        correctQuestions: [
          {question: "Test Question 1"},
          {question: "Test Question 2"}
        ],
        incorrectQuestions: [
          {question: "Test Question 3"},
          {question: "Test Question 4"}
        ]
      });
    });

    it('should combine both correct and incorrect questions and assign them to currentQuestions except for the first value', () => {
      let combinedQuestions = wrapper.state().correctQuestions.concat(wrapper.state().incorrectQuestions).slice(1);
      wrapper.instance().resetAllQuestions();
      expect(wrapper.state().currentQuestions).toEqual(combinedQuestions);
    });

    it('should assign the first value of the combined questions to currentQuestion', () => {
      let firstValue = wrapper.state().correctQuestions.concat(wrapper.state().incorrectQuestions)[0];
      wrapper.instance().resetAllQuestions();
      expect(wrapper.state().currentQuestion).toEqual(firstValue);
    });

    it('should set the correct and incorrect questions to empty arrays', () => {
      wrapper.instance().resetAllQuestions();
      expect(wrapper.state().correctQuestions).toEqual([]);
      expect(wrapper.state().incorrectQuestions).toEqual([]);
    });
  });

  describe('resetIncorrectQuestions', () => {
    beforeEach( () => {
      wrapper.setState({
        incorrectQuestions: [
          {question: "Test Question 1"},
          {question: "Test Question 2"},
          {question: "Test Question 3"},
          {question: "Test Question 4"}
        ]
      });
    });

    it('should set the currentQuestions to the all of the incorrect questions minus the first incorrect question', () => {
      let expectedQuestions = Object.assign([], wrapper.state().incorrectQuestions).splice(1);
      wrapper.instance().resetIncorrectQuestions();
      expect(wrapper.state().currentQuestions).toEqual(expectedQuestions);
    });
    
    it('should set the currentQuestion to the first incorrect question', () => {
      let firstIncorrectQuestion = wrapper.state().incorrectQuestions[0];
      wrapper.instance().resetIncorrectQuestions();
      expect(wrapper.state().currentQuestion).toEqual(firstIncorrectQuestion);
    });

    it('should set the incorrect questions array to an empty array', () => {
      wrapper.instance().resetIncorrectQuestions();
      expect(wrapper.state().incorrectQuestions).toEqual([]);
    });
  });
});
