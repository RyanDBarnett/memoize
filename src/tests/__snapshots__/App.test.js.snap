// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with all data passed in 1`] = `
<div
  className="App"
>
  <img
    className="mandala"
    src="blue-mandala.png"
  />
  <header
    className="App-header"
  >
    <h1>
      ARIA Flashcards
    </h1>
  </header>
</div>
`;
